### ISSUE TITLE
Make markup-safe rendering in Mako the default

### ISSUE THREAD

**Issue by abought** [CenterForOpenScience/osf.io#2933]
## Purpose
Presently, Mako does not escape strings by default. Turning this feature on could simplify our current array of sanitizing/escaping functions, and more reliably guard against potential script injection exploits.

To avoid an excessive QA burden, this proposal would migrate views over individually. We would start with public-facing pages, as discussed with @JeffSpies .

## Time estimate and potential pitfalls
Estimate 1 day of work to migrate wiki and profile pages, and update unit tests. Up to another day for code review responses + QA once merged.

We can take advantage of existing groundwork to change files. Half the time estimate is for validation: ensuring that `mod-meta` and `include`d templates are changed, and reviewing points where markup changes might affect variables passed into JS. (`contextVars`, `base.mako`, etc)

## To-do list

A checklist for template and view changes is available on request; contact @abought for link.

- [ ] Convert routes and pages to work with markupsafe rendering.
    - [x] Infrastructure (common settings and changes to be used by all templates), #2917
    - [x] Profile and settings pages #3635
    - [x] Projects (core features) #3659
        - [ ] Wiki (and other addons)
    - [ ] Misc public pages (landing, login/logout, poster service, search, etc)
    - [ ] Emails
    - [ ] Change default `trust` settings of OsfWebRenderer and `push_status_message` to `False`, so that all new pages will be held to this standard from the start.
- [ ] Audit all javascript views for unsafe uses of text/ markup
- [ ] Evaluate and consolidate sanitization: bleach library adds undesired characters to DB and adds complexity to codebase.
- [ ] After all views are safe by default, and sanitization consolidated: Run DB migration script that corrects escaped HTML data in the database. #3753

## References and previous work
- Change list relies heavily on work by @lyndsysimon in #1308
- Incrementally migrating views to markupsafe relies on #2917 (changes renderer and base templates)
- Initial work on profile page migration #2925 
- HTML entity migration script: : https://github.com/CenterForOpenScience/osf.io/blob/30d3b46a2f0030672378ce84b5cc7accae3cb7ad/scripts/consistency/fix_html_entities.py

**Comment by Ghalko** [comment]
Is this assigned? It looks like it is being worked on. @lbanner

**Comment by lbanner** [comment]
Check with @abought I think -

**Comment by abought** [comment]
I've been working on this but cabn't assing myself.  I've gotten basic utils (#2917) and profile conversion (#2933), as well as initial QA on project page conversion. Haven't pushed too hard given the need to focus on 0.38 and 0.39, but the pieces are here. Suggestions welcome.

After phase 1 (mako) is completed, the big challenge will be phase 2-3 (JS widgets + consolidating backend sanitizers). I probably need to write the latter phases up as a formal proposal; issue is that our JS code may currently be treating data as safe, which would be bad in the sanitize-on-display model being proposed.

**Comment by Ghalko** [comment]
@abought My first suggestion (and possibly only for this one) is to ask @lbanner or @GaryKriebel to add you to your own proposals as the assignee.

**Comment by lbanner** [comment]
@abought -- I assigned you to this one.

**Comment by abought** [comment]
While updating profile PR, I wrote down the checklist of most common things converted. Checklist below.

```
‘Give a man a fish and feed him for a day. Don’t teach a man to fish…and feed yourself. He’s a grown man. And fishing’s not that hard’
— Ron Swanson
```

### Things to examine while converting a given route
- Check every parent template (may be several layers):   `<%inherit file="base.mako"/>`
- Check every included template (may be several layers): `<%include file="include/profile/schools.mako" />`
- Check every mod-meta template (depending on the route, this can be many layers deep) `<div mod-meta='`
- Check that every variable used inside a `<script type="text/javascript”>` tag has the right filters, eg `${ python_var | sjson, n }`
- Generally examine other mako variables to see if anything else might break. Pay particular attention to places where strings are constructed in HTML attributes: 
- In selected instances, the `sjson,n` pair may also be needed for data passed into mod-meta as kwargs. Verify that included/mod-meta templates work as intended.
- Check the associated view functions (and their nested children) for any usages of `framework.status.push_status_message`. Ensure that those are flagged for the appropriate markup option (`trust=True` if they contain HTML)
- Audit any error messages raised by the view, and see if the `message_short` / `message_long` data is injecting HTML directly. (as used in `error.mako`)


### Anti-patterns
- We should never build javascript up inside strings. If a mako variable appears inside a script tag, remove the outer apostrphes and add correct filters, so that ’${ python_var }’ becomes ${ python_var | sjson, n }
- The js_str filter is deprecated. Remove it with fire and sword; use sjson,n instead.
- Additional HTML unescaping should not be done inside the template. That is a symptom of a larger problem. Javascript frameworks are responsible for sanitizing the data they display- whether we pass unsafe data to JS should be a single consistent decision, not something decided per variable.
- Look carefully at HTML attributes that are built up as strings. Some are ok (particularly if they pass server-side validation), but others could be a way to insert code. Will the page render oddly if the URL includes a quotation mark? What about a URL of “javascript:alert(‘Hello World’);”  ?    <a href="${profile['url']}">


Validation
Only convert one route at a time. Then load that page and examine it for oddities. Contact abought for access to a private gist containing some useful test strings; suggestions are welcomed!

Run unit tests. All should pass.

Stay aware of errors, but keep code quality high during the conversion. Mark down things that we SHOULD fix in this template, and things we SHOULDN’T. (eg if bad data is stored in the database, don’t try to hack around it in a template. That will help short term, but eventually templates would be filled with ugly hacks that were no longer relevant. 

The javascript console is a common place for errors, especially in JS strings after the conversion. Read it carefully on every page you check.

Some variables won’t get triggered locally; check carefully in both local and production environments.
